### PR TITLE
Allow `source` in hooks to point to relative paths

### LINKED ISSUE #1557: [neomuttrc] Cannot source file from folder-hook using relative path
When try to set up a folder-hook to source a file from `~/.config/neomutt/`, the hook fails with the following error:
```
Error: issue converting path to absolute (No such file or directory)
```
And the debug log shows:
```
[2019-02-10 20:45:15]<E> source_rc() Error: impossible to build path of 'account.mail'.
[2019-02-10 20:45:15]<E> mutt_folder_hook() source: file account.mail could not be sourced.
```
To reproduce, I used the following minimal test case:
**~/.config/neomutt/test.rc**
```
folder-hook account 'source account.mail'
set mbox_type=Maildir
set folder="~/mail"
mailboxes = `find ~/mail -mindepth 1 -type d ! -name cur ! -name tmp ! -name new -printf '"%p"\0' | xargs -0`
source account.mail
```
**~/.config/neomutt/account.mail**
```
set spoolfile = "+account/Inbox"
```
When I run `neomutt -F ~/.config/neomutt/test.rc` the **account.mail** file gets sourced correctly the first time by the bare source call, but then the folder-hook is triggered and fails.
This setup works correctly when I give an absolute path to the folder-hook command.

### PR DESCRIPTION
**Pull request by mjsir911**
Fixes #1557

### CODE REVIEW

File: command_parse.h [hunk 1]
```diff
@@ -65,6 +65,9 @@ enum CommandResult parse_unsubscribe     (struct Buffer *buf, struct Buffer *s,
 enum CommandResult parse_unsubscribe_from(struct Buffer *buf, struct Buffer *s, intptr_t data, struct Buffer *err);
 #endif
 
+enum CommandResult mutt_parse_rc_line_cwd(const char *line, struct Buffer *err, char *cwd);
```
Comment (mjsir911) [inline command_parse.h:68]: eh should the arg order be re-ordered to put `err` at the end?

File: command_parse.h [hunk 2]
```diff
@@ -65,6 +65,9 @@ enum CommandResult parse_unsubscribe     (struct Buffer *buf, struct Buffer *s,
...
+char *mutt_get_sourced_cwd(void);
```
Comment (mjsir911) [inline command_parse.h:69]: is (void) necessary? I think it flags argument-less function rather than unknown argd function?

File: hook.c [hunk 3]
```diff
@@ -281,6 +283,7 @@ enum CommandResult mutt_parse_hook(struct Buffer *buf, struct Buffer *s,
          * based upon some other information. */
         FREE(&hook->command);
         hook->command = mutt_buffer_strdup(cmd);
+        hook->cwd = mutt_get_sourced_cwd(); // ?
```
Comment (mjsir911) [inline hook.c:286]: not sure what this code path is doing, what kind of hook is it?

**Comment by mjsir911** [comment]
Oh, related. This invalidates some of what #3470 is saying @whitney-cucumber

**Comment by mjsir911** [comment]
Another weird behavior that might come up is what to do about hooks specified at runtime, outlined in https://github.com/neomutt/neomutt/issues/1557#issuecomment-477991396. 

Current behavior if no relative file is found is fallbacking on `pwd`, which happens at hook-calltime rather than hook-setuptime. I can't really imagine a situation where neomutt's `cwd` changes on-the-go, but the fix would be pretty easy:

```diff
-         hook->cwd = mutt_get_sourced_cwd(); // ?
+         hook->cwd = mutt_get_sourced_cwd() || getcwd(); // ?

**Comment by mjsir911** [comment]
Another weird behavior that might come up is what to do about hooks specified at runtime, outlined in https://github.com/neomutt/neomutt/issues/1557#issuecomment-477991396. 

Current behavior if no relative file is found is fallbacking on `pwd`, which happens at hook-calltime rather than hook-setuptime. I can't really imagine a situation where neomutt's `cwd` changes on-the-go, but the fix would be pretty easy:

```diff
-         hook->cwd = mutt_get_sourced_cwd(); // ?
+         hook->cwd = mutt_get_sourced_cwd() || getcwd(); // ?
```

File: command_parse.h (hunk 1 above)
Comment (flatcap) [reply command_parse.h:68]: yes please

File: command_parse.h (hunk 2 above)
Comment (flatcap) [reply command_parse.h:69]: Yes.  in C++ `function()` specifically means no parameters.
In C, it means unknown, which means the compiler can't issue warnings.

File: hook.c [hunk 4]
```diff
@@ -64,6 +65,7 @@ struct Hook
   HookFlags type;              ///< Hook type
   struct Regex regex;          ///< Regular expression
   char *command;               ///< Filename, command or pattern to execute
+  char *cwd;                   ///< Used for relative-directory source
```
Comment (flatcap) [inline hook.c:68]: I think a better name might be something like:
```c
source_file; ///< Config file that defined the hook
```

File: command_parse.c [hunk 5]
```diff
@@ -135,6 +135,24 @@ int parse_grouplist(struct GroupList *gl, struct Buffer *buf, struct Buffer *s,
...
+    STAILQ_REMOVE_HEAD(&MuttrcStack, entries);
+    FREE(&np->data);
+    FREE(&np);
+
+    return ret;
+}
+
+char *mutt_get_sourced_cwd(void) { // caller is responsible for freeing returned string
+    struct ListNode *np = STAILQ_FIRST(&MuttrcStack);
+    return np ? np->data ? mutt_str_dup(np->data) : NULL : NULL;
```
Comment (flatcap) [inline command_parse.c:153]: If the `MuttrcStack` is empty, then we should add the current working dir.
My quick hack was:
```c
  struct Buffer *cwd = mutt_buffer_pool_get();
  mutt_path_getcwd(cwd);
  mutt_buffer_addstr(cwd, "/dummy.rc");
  char *dir = mutt_buffer_strdup(cwd);
  mutt_buffer_pool_release(&cwd);
  return dir;
```

File: hook.c [hunk 6]
```diff
@@ -322,6 +325,7 @@ enum CommandResult mutt_parse_hook(struct Buffer *buf, struct Buffer *s,
   hook = mutt_mem_calloc(1, sizeof(struct Hook));
   hook->type = data;
   hook->command = mutt_buffer_strdup(cmd);
+  hook->cwd = mutt_get_sourced_cwd();
```
Comment (flatcap) [inline hook.c:328]: These need to be freed in `delete_hook()`

**Review by flatcap — CHANGES_REQUESTED** [review]
OK, it looks like it's nearly there.

Once you're done with the changes, please can you:
- Add Doxygen comment blocks for your new functions
  https://neomutt.org/dev/doxygen
- Run `clang-format` on your changes
  https://neomutt.org/dev/analysis/clang-format
  https://neomutt.org/dev/code
- Rebase the code onto `main`

Thanks

**Review by flatcap — APPROVED** [review]
Great job!  Thanks

**Pull request merged**